src/tensor_ops/reshape.rs

```diff
@@ -0,0 +1,68 @@
+use crate::prelude::*;
+
+pub fn reshape<T, R>(t: T) -> R
+where
+    T: Tensor<Dtype = f32>,
+    R: Tensor<Tape = T::Tape, Dtype = f32>,
+    ConstEq<{ T::Array::NUM_ELEMENTS }, { R::Array::NUM_ELEMENTS }>: ConstAssert<Result = True>,
+    ConstEq<{ R::Array::NUM_ELEMENTS }, { T::Array::NUM_ELEMENTS }>: ConstAssert<Result = True>,
+{
+    let (mut t, mut tape) = t.split_tape();
+    let mut result: R::NoTape = R::NoTape::zeros();
+    copy(t.data(), result.mut_data());
+    let _result = result.phantom();
+    tape.add_backward_op(move |grads| {
+        let result_grad: &R::Array = grads.ref_gradient(&_result);
+        copy(result_grad, t.mut_data());
+        T::Device::add(grads.mut_gradient(&t), t.data());
+    });
+    result.put_tape(tape)
+}
+
+fn copy<Lhs: CountElements, Rhs: CountElements<Dtype = Lhs::Dtype>>(lhs: &Lhs, rhs: &mut Rhs)
+where
+    ConstEq<{ Lhs::NUM_ELEMENTS }, { Rhs::NUM_ELEMENTS }>: ConstAssert<Result = True>,
+{
+    let l = lhs.ref_first_elem() as *const Lhs::Dtype;
+    let r = rhs.mut_first_elem() as *mut Lhs::Dtype;
+    unsafe {
+        std::ptr::copy_nonoverlapping(l, r, Lhs::NUM_ELEMENTS);
+    }
+}
+
+pub struct True;
+pub trait ConstAssert {
+    type Result;
+}
+
+pub struct ConstEq<const A: usize, const B: usize>;
+impl<const N: usize> ConstAssert for ConstEq<N, N> {
+    type Result = True;
+}
```

Comment:
I think a slightly cleaner set of traits would be the following:
```rust
pub trait ConstTrue{}
pub struct ConstEq<const A: usize, const B: usize>;

impl<const N: usize> ConstTrue for ConstEq<N, N> {}
```

This would change your where clause above from this:
```rust
ConstEq<{ T::Array::NUM_ELEMENTS }, { R::Array::NUM_ELEMENTS }>: ConstAssert<Result = True>
```

To this:
```rust
ConstEq<{ T::Array::NUM_ELEMENTS }, { R::Array::NUM_ELEMENTS }>: ConstTrue
```

Reply:
Oooh this is great thanks!